### DISCUSSION TITLE
Data on respondent's technical background

### DISCUSSION THREAD

**Discussion by LeaVerou** [Devographics/surveys#168] [State of HTML Design]
Like CSS, HTML is written by many types of authors, with different backgrounds. In fact, I suspect there is even bigger variance in this for HTML, as it is the most approachable of the web platform languages.

In State of CSS, we have this question:

<img width="816" alt="image" src="https://github.com/Devographics/surveys/assets/175836/dd992ee3-1d25-4554-a556-24f7cf1413c0">

We could either have a similar question (obviously changing "CSS" with "HTML" in the title), or break it down further with something like: 

> In 10 typical hours of your typical work with web technologies, how many are…
> 
> HTML: ____
> CSS: ____
> JS: ____
> Backend: ____
> Non-coding: ____

(with JS to enforce the sum being 10)

This gives us a much better picture, at the obvious downside that it's significantly more effort to fill in as it's annoying to manually balance the sum.

Alternatively, we could have independent sliders that adjust percentages. I made an interactive prototype [here](https://codepen.io/leaverou/pen/KKrEvVj) and here’s a video of the interaction:

https://github.com/Devographics/surveys/assets/175836/8a0c451e-7482-461f-801d-59d0f6b326b0

Obviously many kinks to iron out, but I think it may work?

**Comment by AnthonyLenglet** [comment]
Could this data be gathered by asking people how much they would say they know about html going in, asking the same question again at the end of the survey, and use the delta of that to get the technical background ?

One potential pitfall of this approach is that it might feel too reductive (basically just throwing really niche information at people and saying "aha yuore wrong"), but it would fix the effort part

**Reply by LeaVerou** [reply]
It's well known that there are big confounds in how people from minorities and underrepresented groups estimate their knowledge in such questions.

**Reply by AnthonyLenglet** [reply]
good point, hadn't really though about that

I think if this question ends up staying as is, it would be good to keep in mind the amount of time spent on a device not actually coding (documentation, testing (not necessarily unit testing), keeping up with tech,...), as it stands, I'm not actually sure how I would personally answer this question based on everything I might switch to randomly

Also worth considering that the answer will vary wildly depending on whether I'm on a new project or debugging an existing one, and I'm not use which one would be considered "typical"

**Comment by ShaineRosewel** [comment]
Adding to 10 is a good idea since it allows us to extract meaningful information.

Alternatively, it could be ranking or ratings. This entails less effort for the respondents since they do not need to think of adding to a whole. Ranking will force the respondents to rank the four according to time spent - keeping the responses strictly dependent. If it is rating then, we could ask them to choose say from 1 to 5, rate according to time/effort it takes with 1 being the least and 5 the most - this allows them to think of four more independently from each other.

Disadvantage is we get less meaningful data as we cannot really quantify the difference, say between 1 and 2 in a ranking. The same holds true for ratings.

**Reply by LeaVerou** [reply]
Ranking will not give us the kind of data we want as it doesn't capture the relative proportion. The rating idea is interesting, I just had an idea about a hybrid of the independent ratings + adding to 10: respondent provides independent ratings, and they are scaled down to percentages which are communicated to the respondent via a pie chart. I'll prepare a mockup later today as I'm not sure how clear this is from a textual description.

**Reply by LeaVerou** [reply]
I just updated the OP with the prototype!

**Comment by SachaG** [comment]
Just commenting on the UI side of things, what if instead of sliders we had +/- buttons, with the + becoming greyed out once you've allocated all your "points"?

**Reply by LeaVerou** [reply]
Could you expand on that? What advantage do you feel it offers over the sliders?

**Reply by SachaG** [reply]
Maybe easier to understand and use than the sliders? Here's a very rough mockup:

![Frame 5](https://github.com/Devographics/surveys/assets/358832/1aef3e32-ab3d-4daa-8dc7-a709429df950)

**Reply by SachaG** [reply]
Or, to make the hours a bit more explicit (but we probably have to limit it to 10 or 12 in that case):

![Frame 6](https://github.com/Devographics/surveys/assets/358832/7a54e25e-d69e-4062-a381-81326e577259)

(when you start out all 12 "hours to allocate" squares would be filled out, and they'd empty out as you allocate the hours in one of the categories)

**Reply by LeaVerou** [reply]
My hypothesis is that people don't think in terms of allocation of 8, 10, or 12 hours, but they do have a sense of what proportion of their time they spend on each thing.
What I'm trying to avoid with the slider idea is respondents having to do mental gymnastics like having to take an hour from CSS and add it to JS to make the proportions seem right. While this minimizes the issue of having to manually make sure things add up to a specific number (though it's still there, just easier to deal with), it still requires mental gymnastics to express a given proportion.
Also, this requires 12 clicks no matter what, whereas with the sliders if you spend your time on 1-2 things, it's fairly quick to answer.

**Reply by SachaG** [reply]
What if we let respondents click on a box to fill up all the boxes up to the selected box? So it would become a hybrid between the sliders and the boxes ideas. 

P.S. Also I'm not really sure I'll have time to implement this new input method for State of HTML, but I do like the concept!

**Reply by Danny-Engelman** [reply]
A slider can be designed: 
https://sipsandbits.com/2020/10/21/create-a-nice-looking-input-range-with-only-css/

**Reply by LeaVerou** [reply]
> P.S. Also I'm not really sure I'll have time to implement this new input method for State of HTML, but I do like the concept!

The slider prototype is actually functional. Obviously some work would be needed to hook it up to the data model and align its code to the codebase, but perhaps less than you may think :)

**Comment by Danny-Engelman** [comment]
We calculate time in time units. Not _I have been to lunch for 11%_

So maybe change the sliders to ``step="15"`` and display the value in **H:MM** notation.

That will also allow DEVs to have 14 hour working days

Pie chart can still be a pie chart: https://jsfiddle.net/WebComponents/d68jebwr/

![](https://i.imgur.com/MkGEBJc.png)

**Reply by LeaVerou** [reply]
Wow, thanks for spending time on this!
We're trying to see what *type* of professional they are, not how much they work. I'd rather rephrase the question to make it clear that procrastination or lunch time is not part of it.
Also, the point is not precision, it's to distinguish e.g. backend developers, JS developers, web designers, PMs etc

**Comment by Danny-Engelman** [comment]
Maybe test your hypothesis.

Not many people understand proportions.

_The firm conducted a focus group and found that around half of the people surveyed thought that the A&W 1/3 pound burger was smaller than McDonald’s 1/4 pounder! “Why should we pay the same amount for a third of a pound of [meat](https://culinarylore.com/food-culture:what-is-meat/) as we do for a quarter-pound of meat?” they said._

Source: https://culinarylore.com/food-history:aw-1-3-pound-burger-failure-fact-check-are-americans-really-that-bad-at-fractions/

**Reply by LeaVerou** [reply]
> Maybe test your hypothesis.

Absolutely! That’s why we plan to do a round of user testing before launch :) 

> Not many people understand proportions.

I’m aware of the issue you’re citing wrt fractions, however do note that:
a) Understanding of fractions and understanding of proportions is not the same thing. Fractions are a way to write proportions. **Understanding a way to serialize a concept in writing is not the same as understanding the concept itself.** To use this example, even someone who wouldn't be able to tell that 1/3 is bigger than 1/4 if they saw them in writing, would still be able to distribute a 12 slice pizza to a group of 3 or 4 people, and would be able to tell that 4 slices (1/3) is more food than 3 slices (1/4). Similarly, when small children are learning numbers, learning to process the actual symbols for numbers is a different skill than actually understanding the quantities visually. 
b) These statistics apply to the general population; developers are not a representative sample of the general population.

**Reply by LeaVerou** [reply]
Although, from these comments I’m starting to wonder if displaying the percentage selected is actually *not* helpful, as it makes people think they need to input percentages. Perhaps simply showing a pie chart with no exact numbers would be better to communicate that exact numbers don't matter here.
